File: graphs/bfs/433-minimum-genetic-mutation/MinimumGeneticMutation.py
class Solution:
    def minMutation(self, start: str, end: str, bank: List[str]) -> int:
        bank = set(bank)
        if end not in bank:
            return -1

        queue = deque([(start, 0)])
        visited = {start}
        genes = ['A', 'C', 'G', 'T']

        while queue:
            gene, steps = queue.popleft()
            if gene == end:
                return steps

            for i in range(len(gene)):
                for g in genes:
                    mutated = gene[:i] + g + gene[i+1:]
                    if mutated in bank and mutated not in visited:
                        visited.add(mutated)
                        queue.append((mutated, steps + 1))
        return -1

        

File: graphs/bfs/433-minimum-genetic-mutation/MinimumGeneticMutation.java
class Solution {
    public int minMutation(String start, String end, String[] bank) {
        Set<String> geneBank = new HashSet<>(Arrays.asList(bank));
        if (!geneBank.contains(end)){
            return -1;
        } 

        char[] genes = {'A', 'C', 'G', 'T'};
        Queue<String> queue = new LinkedList<>();
        queue.offer(start);
        Set<String> visited = new HashSet<>();
        visited.add(start);

        int level = 0;
        while (!queue.isEmpty()) {
            int size = queue.size();
            for (int i = 0; i < size; i++) {
                String curr = queue.poll();
                if (curr.equals(end)) return level;

                char[] arr = curr.toCharArray();
                for (int j = 0; j < arr.length; j++) {
                    char old = arr[j];
                    for (char g : genes) {
                        arr[j] = g;
                        String next = new String(arr);
                        if (geneBank.contains(next) && !visited.contains(next)) {
                            visited.add(next);
                            queue.offer(next);
                        }
                    }
                    arr[j] = old;
                }
            }
            level++;
        }
        return -1;
        
    }
}